src/workerd/api/filesystem.c++

```diff
@@ -2019,17 +2019,106 @@ jsg::Ref<jsg::DOMException> fsErrorToDomException(jsg::Lock& js, workerd::FsErro
 }
 }  // namespace
 
+FileSystemHandle::FileSystemHandle(
+    const workerd::VirtualFileSystem& vfs, jsg::Url&& locator, jsg::USVString name)
+    : vfs(vfs),
+      locator(kj::mv(locator)),
+      name(kj::mv(name)) {}
+
+jsg::Promise<kj::StringPtr> FileSystemHandle::getUniqueId(
+    jsg::Lock& js, const jsg::TypeHandler<jsg::Ref<jsg::DOMException>>& deHandler) {
+  KJ_IF_SOME(item, vfs.resolve(js, getLocator(), {})) {
+    KJ_SWITCH_ONEOF(item) {
+      KJ_CASE_ONEOF(file, kj::Rc<workerd::File>) {
+        return js.resolvedPromise(file->getUniqueId(js));
+      }
+      KJ_CASE_ONEOF(dir, kj::Rc<workerd::Directory>) {
+        return js.resolvedPromise(dir->getUniqueId(js));
+      }
+      KJ_CASE_ONEOF(link, kj::Rc<workerd::SymbolicLink>) {
+        return js.resolvedPromise(link->getUniqueId(js));
+      }
+      KJ_CASE_ONEOF(err, workerd::FsError) {
+        return js.rejectedPromise<kj::StringPtr>(
+            deHandler.wrap(js, fsErrorToDomException(js, err)));
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+  auto ex = js.domException(kj::str("NotFoundError"), kj::str("The entry was not found."));
+  return js.rejectedPromise<kj::StringPtr>(deHandler.wrap(js, kj::mv(ex)));
+}
+
+jsg::Promise<bool> FileSystemHandle::isSameEntry(jsg::Lock& js, jsg::Ref<FileSystemHandle> other) {
+  // Per the spec, two handles are the same if they refer to the same entry (that is,
+  // have the same locator). It does not matter if they are different actual entries.
+  return getKind(js) == other->getKind(js) &&
+          locator.equal(other->getLocator(),
+              jsg::Url::EquivalenceOption::IGNORE_FRAGMENTS |
+                  jsg::Url::EquivalenceOption::IGNORE_SEARCH |
+                  jsg::Url::EquivalenceOption::NORMALIZE_PATH)
+      ? js.resolvedPromise(true)
+      : js.resolvedPromise(false);
+}
+
+jsg::Promise<void> FileSystemHandle::remove(jsg::Lock& js,
+    jsg::Optional<RemoveOptions> options,
+    const jsg::TypeHandler<jsg::Ref<jsg::DOMException>>& deHandler) {
+  auto relative = getLocator().getRelative();
+  auto opts = options.orDefault(RemoveOptions{});
+  auto recursive = opts.recursive.orDefault(false);
+  KJ_IF_SOME(parent, vfs.resolve(js, relative.base, workerd::VirtualFileSystem::ResolveOptions{})) {
+    KJ_SWITCH_ONEOF(parent) {
+      KJ_CASE_ONEOF(parentDir, kj::Rc<workerd::Directory>) {
+        // Webfs requires that the entry exists before we try to remove it.
+        kj::Path path({name});
+        if (parentDir->stat(js, path) == kj::none) {
+          auto ex = js.domException(kj::str("NotFoundError"), kj::str("The entry was not found."));
+          return js.rejectedPromise<void>(deHandler.wrap(js, kj::mv(ex)));
+        }
```

Comment:
This looks the same as: https://github.com/cloudflare/workerd/pull/4584/files#diff-f82d13fe2be1bc6edb903a4f830ad17b7209d8da05cb7b45eb1e8e224cbc96beR2085

I'm not sure I understand the distinction between the two cases.

Reply:
Just two different paths to ending at the same result.